src/util.cc

```diff
@@ -239,22 +241,38 @@ bool CanonicalizePath(char* path, size_t* len, unsigned int* slash_bits,
   return true;
 }
 
-static inline bool IsKnownShellSafeCharacter(char ch) {
-  if ('A' <= ch && ch <= 'Z') return true;
-  if ('a' <= ch && ch <= 'z') return true;
-  if ('0' <= ch && ch <= '9') return true;
+struct IsKnownShellSafeCharacterImpl {
+  IsKnownShellSafeCharacterImpl() {
+    for (int ch = 0; ch < UCHAR_MAX; ch++) {
+      known_shell_safe_characters_[ch] = safe(ch);
+    }
+  }
 
-  switch (ch) {
-    case '_':
-    case '+':
-    case '-':
-    case '.':
-    case '/':
-      return true;
-    default:
-      return false;
+  bool safe(char ch) {
+    if ('A' <= ch && ch <= 'Z') return true;
+    if ('a' <= ch && ch <= 'z') return true;
+    if ('0' <= ch && ch <= '9') return true;
+
+    switch (ch) {
+      case '_':
+      case '+':
+      case '-':
+      case '.':
+      case '/':
+        return true;
+      default:
+        return false;
+    }
   }
-}
+
+  bool operator()(char ch) const {
+    return known_shell_safe_characters_[static_cast<unsigned char>(ch)];
+  }
+
+  bool known_shell_safe_characters_[UCHAR_MAX];
```

Comment:
Can't this be made static, and then initialized once at program start?

Better even would be to conditionally compile so that when C++11 is available, the initialization of this array is constexpr, and the array is part of the read-only-data section of the binary.